src/sinks/clickhouse/service.rs

```diff
@@ -0,0 +1,252 @@
+use http::{
+    header::{CONTENT_ENCODING, CONTENT_LENGTH, CONTENT_TYPE},
+    Request, Response, StatusCode, Uri,
+};
+use hyper::{body, Body};
+use snafu::ResultExt;
+use std::task::{Context, Poll};
+use tracing::Instrument;
+
+use crate::{
+    http::{Auth, HttpClient, HttpError},
+    sinks::{
+        prelude::*,
+        util::{http::HttpRetryLogic, retries::RetryAction},
+        UriParseSnafu,
+    },
+};
+
+#[derive(Debug, Clone)]
+pub struct ClickhouseRequest {
+    pub body: Bytes,
+    pub compression: Compression,
+    pub finalizers: EventFinalizers,
+    pub metadata: RequestMetadata,
+}
+
+impl MetaDescriptive for ClickhouseRequest {
+    fn get_metadata(&self) -> RequestMetadata {
+        self.metadata
+    }
+}
+
+impl Finalizable for ClickhouseRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        self.finalizers.take_finalizers()
+    }
+}
+
+pub struct ClickhouseResponse {
+    metadata: RequestMetadata,
+    http_response: Response<Bytes>,
+}
+
+impl DriverResponse for ClickhouseResponse {
+    fn event_status(&self) -> EventStatus {
+        match self.http_response.status().is_success() {
+            true => EventStatus::Delivered,
+            false => EventStatus::Rejected,
+        }
+    }
+
+    fn events_sent(&self) -> CountByteSize {
+        CountByteSize(
+            self.metadata.event_count(),
+            self.metadata.events_estimated_json_encoded_byte_size(),
+        )
+    }
+
+    fn bytes_sent(&self) -> Option<usize> {
+        Some(self.metadata.request_encoded_size())
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct ClickhouseRetryLogic {
+    inner: HttpRetryLogic,
+}
+
+impl RetryLogic for ClickhouseRetryLogic {
+    type Error = HttpError;
+    type Response = ClickhouseResponse;
+
+    fn is_retriable_error(&self, error: &Self::Error) -> bool {
+        self.inner.is_retriable_error(error)
+    }
+
+    fn should_retry_response(&self, response: &Self::Response) -> RetryAction {
+        match response.http_response.status() {
+            StatusCode::INTERNAL_SERVER_ERROR => {
+                let body = response.http_response.body();
+
+                // Currently, ClickHouse returns 500's incorrect data and type mismatch errors.
+                // This attempts to check if the body starts with `Code: {code_num}` and to not
+                // retry those errors.
+                //
+                // Reference: https://github.com/vectordotdev/vector/pull/693#issuecomment-517332654
+                // Error code definitions: https://github.com/ClickHouse/ClickHouse/blob/master/dbms/src/Common/ErrorCodes.cpp
+                //
+                // Fix already merged: https://github.com/ClickHouse/ClickHouse/pull/6271
+                if body.starts_with(b"Code: 117") {
+                    RetryAction::DontRetry("incorrect data".into())
+                } else if body.starts_with(b"Code: 53") {
+                    RetryAction::DontRetry("type mismatch".into())
```

Comment:
Is this still applicable after all this time?

Reply:
Good question, I thought about this. I considered removing it to simplify the code. However, I opted against it because we already had it and it would change behavior for anyone still running Clickhouse before that version. Now, the likelihood of that being the case is extremely low, so I'd be happy to remove it if others agree.

Reply:
I'm frequently trying to pull in to many changes in a single PR, even if they're unrelated. I'd be fine leaving it for another task, a new TODO/issue to track, etc.

It's definitely not in scope for the original "refactor to stream" work.